src/loaders.py

```diff
@@ -31,69 +32,33 @@ def __init__(self, X, y, train_mode,
         self.image_augment = image_augment
         self.image_augment_with_target = image_augment_with_target
 
-    def load_image(self, img_filepath):
-        image = Image.open(img_filepath, 'r')
-        return image.convert('RGB')
+        self.image_source = image_source
 
     def __len__(self):
-        return self.X.shape[0]
+        if self.image_source == 'memory':
+            return len(self.X[0][0])
+        elif self.image_source == 'disk':
+            return self.X.shape[0]
 
     def __getitem__(self, index):
-        img_filepath = self.X[index]
-        Xi = self.load_image(img_filepath)
-
-        if self.y is not None:
-            mask_filepath = self.y[index]
-            Mi = self.load_image(mask_filepath)
-
-            if self.train_mode and self.image_augment_with_target is not None:
-                Xi, Mi = from_pil(Xi, Mi)
-                Xi, Mi = self.image_augment_with_target(Xi, Mi)
-                Xi = self.image_augment(Xi)
-                Xi, Mi = to_pil(Xi, Mi)
-
-            if self.mask_transform is not None:
-                Mi = self.mask_transform(Mi)
-
-            if self.image_transform is not None:
-                Xi = self.image_transform(Xi)
-            return Xi, Mi
+        if self.image_source == 'memory':
+            load_func = self.load_from_memory
+        elif self.image_source == 'disk':
+            load_func = self.load_from_disk
         else:
-            if self.image_transform is not None:
-                Xi = self.image_transform(Xi)
-            return Xi
-
-
-class ImageSegmentationDataset(Dataset):
-    def __init__(self, X, y, train_mode,
-                 image_transform, image_augment_with_target,
-                 mask_transform, image_augment):
-        super().__init__()
-        self.X = X
-        if y is not None:
-            self.y = y
-        else:
-            self.y = None
-
-        self.train_mode = train_mode
-        self.image_transform = image_transform
-        self.mask_transform = mask_transform
-        self.image_augment = image_augment
-        self.image_augment_with_target = image_augment_with_target
+            raise NotImplementedError("Possible loading options: 'memory' and 'disk'!")
 
-    def __len__(self):
-        return len(self.X[0])
-
-    def __getitem__(self, index):
-        Xi = self.X[0][index]
+        Xi = load_func(self.X, index)
 
         if self.y is not None:
-            Mi = self.y[0][index]
+            Mi = load_func(self.y, index)
 
-            if self.train_mode and self.image_augment_with_target is not None:
+            if self.image_augment_with_target is not None or self.image_augment is not None:
```

Comment:
@apyskir this looks weird. We are checking condition 1 and 2 and inside of that we are checking them again one by one ?

Reply:
@jakubczakon Yes, if any of the augmentations is not None, we want to transform to PIL, and then we want to perform the augmentations if they are not None. I can propose, that I will write:
```
if self.augment_image is None:
   self.augment_image = iaa.Noop()
```
This way we can remove checking the conditions inside __get_item__

Reply:
@apyskir Oh I don't mind checking them in __get_item__ i just think that something unnatural is happening here. For example say we want to just augment_image why do we produce Xi, Mi  and not just Xi ?

Reply:
@jakubczakon Because if you want to augment_image_with target separately, then you have to convert Xi to PIL twice. I think it's the best option to use my proposal and then always convert Xi, Mi to PIL, apply augmentations, and then back to numpy. An alternative is:
```if self.augment_image_with_target is not None:
  Xi, Mi = to_pil(Xi, Mi)
  Xi, Mi = self.augment_image_with_target
  Xi, Mi = from_pil(Xi, Mi)
if self.augment_image is not None:
  Xi = to_pil(Xi)
  Xi = self.augment_image
  Xi = from_pil(Xi)```
Am I right? Or I can check all combinations of self.augment_image and self.augment_image_with_target being `None` and `not None` but I don't think it's necessary.

Reply:
@apyskir I like this more, because it is clear what is happening